File: closestpair/closestpair.nim
let a = [-1,3,8,2,9,5]
let b = [4,1,2,10,5,20]
let t = 24 # target

proc pairs(): array[a.len * b.len, (int,int)] =
 var n = 0
 for x in a:
  for y in b:
   result[n] = (x,y)
   inc n

proc fmin[N](arr: array[N,(int,int)]): seq[(int,int)] =
 var min = high(int)
 for p in arr:
  let s = abs(t - (p[0] + p[1]))
  if s < min:
   min = s
   result = @[p]
  elif s == min:
   result.add(p)

for p in fmin pairs():
 echo p

File: closestpair/closestpair.hs
import Control.Arrow

a = [-1,3,8,2,9,5]
b = [4,1,2,10,5,20]
t = 24 -- TARGET

main = print solution
 where
  solution = fmin $ (\x y -> ((x,y),abs$t-(x+y))) <$> a <*> b

fmin :: [(a,Int)] -> [a]
fmin = go [] maxBound
 where
  go acc _ [] = acc
  go acc n ((a,x):as)
   | x <  n = go (a:[] ) x as
   | x == n = go (a:acc) n as
   | True   = go    acc  n as
